### PR TITLE
Helping contributors (new rules regarding white space)

### PR DESCRIPTION
**Pull request by lemire**
It may not be obvious to some why some white-space tests are failing. I believe we need to document it.

### CODE REVIEW

**Comment by lemire** [comment]
@pauldreik Can you have a look?

I am concerned with people submitting a PR and not understand what is happening.

**Comment by friendlyanon** [comment]
It's possible to leave a comment on PRs via Github workflows.  
I think [unsplash/comment-on-pr](https://github.com/unsplash/comment-on-pr) would be a nice way to leave a message if there was a failure to tell the contributor where to find the artifact/error message and how to proceed. It even handles not sending the same message more than once.

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+
+
+Our tests check whether you have introduced trailing white space. If such a test fails, please check the "artifacts button" above, which if you click it gives a link to a downloadable file to help you identify the issue. One can also run scripts/remove_trailing_whitespace.sh locally if you have a bash shell.
```
Comment (pauldreik) [inline .github/pull_request_template.md:3]: Once can also -> You can also

bash shell. -> bash shell and sed available.

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+
+In this case, I think you just needed to merge master into the branch since this branch probably started before I un-whitespaced master.
```
Comment (pauldreik) [inline .github/pull_request_template.md:5]: Remove

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -42,6 +42,7 @@ We have few hard rules, but we have some:
 - Calls to `abort()` are forbidden in the core library. This follows from the [Writing R Extensions](https://cran.r-project.org/doc/manuals/R-exts.html) manual which states that "Under no circumstances should your compiled code ever call abort or exit".
 - All source code files (.h, .cpp) must be ASCII.
 - All C macros introduced in public headers need to be prefixed with either `SIMDJSON_` or `simdjson_`.
+- We avoid trailing white space characters. Please avoid making unnecessary changes to white-space characters.
```
Comment (pauldreik) [inline CONTRIBUTING.md:45]: I would say that trailing white space and formatting are two separate issues. 
I would like this to be only:
- We avoid trailing white space characters.

I agree with trying to keep out of whitespace changes, but that should be covered by some kind of formatting guideline. I plan to have a look on formatting later. @jkeiser said it had been discussed earlier, there is even .clang-format and a CI job using it already. I want to see what has been discussed before I start digging.

**Review by pauldreik — COMMENTED** [review]
> I think [unsplash/comment-on-pr](https://github.com/unsplash/comment-on-pr) would be a nice way to leave a message if there was a failure to tell the contributor where to find the artifact/error message and how to proceed. It even handles not sending the same message more than once.

Having a github message is much nicer!

I tried to understand how to only leave a message in case there is something to do, but could not find out if that is possible. I don't want to contribute to warning fatigue, I just want to see those that require action.

Also, what happens for subsequent pushed? Do you get one message each time?

An alternative would be for the existing github action to somehow print it's error in big red characters, since even @lemire missed the error message :-)

**Comment by friendlyanon** [comment]
> I tried to understand how to only leave a message in case there is something to do

> Do you get one message each time?

You can have a conditional step at the end of the workflow with the appropriate config:
```yml
      - name: Leave a message with instructions
        uses: unsplash/comment-on-pr@v1
        if: failure()
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          msg: "..."
          check_for_duplicate_msg: true
```

**Comment by lemire** [comment]
We do have clang-format conventions and support....

But making something like clang-format compulsory is problematic because 

1. Not everyone has clang-format, and in some cases it is difficult to install (keep in mind that we support a wide range of platforms, not just linux). Put enough obstacles in the way and you will prevent sane people from ever contributing. 
2. As far as I know, there is no guarantee that two different clang-format versions will always generate the same output.


So it is one thing to forbid trailing white spaces... It is something everyone can support... but once you start enforcing styles in C++... in an open project like this one where contributors using different OSes, different compilers, different everything... you really get into an authoritarian regime.

For some projects, this makes sense. If you build a Linux database system, you might enforce strict guidelines... but this library here is meant to be used by a wide range of different folks, and we want a wide range of different contributors.

Also, we have simply *not* have the "here is my favorite style" problem in a long time. We don't have the bikeshedding that occurs in some other projects. There was a few attempts early on to just reformat all of the code, but we made it an explicit policy that we did not want these kind of contributions... and people have understood.

To put it another way, we have managed not to spend a lot of time on style issues, and I submit to you that it is a good thing. You want to work toward features, not style.

File: .github/pull_request_template.md (hunk 2 above)
Comment (lemire) [reply .github/pull_request_template.md:5]: Yep.

File: .github/pull_request_template.md (hunk 1 above)
Comment (lemire) [reply .github/pull_request_template.md:3]: Fixed.

File: CONTRIBUTING.md (hunk 3 above)
Comment (lemire) [reply CONTRIBUTING.md:45]: @pauldreik See me comment later about how we do not want users to have a particular setup, a particular version of clang-format and so forth... to work on simdjson. In Go, for example, we typically require users to do "go fmt" because there is a standard formatter that everyone who uses Go has... But in C++, there is no such thing. That is, a given team might have a strict style policy and strict style guidelines... 

I guess that the equivalent here would be for CMake to provide style checking. If it did so, then we could/should enable it.

**Comment by lemire** [comment]
Note: I am not at all against style checks and constraints. In fact, I have be known to enforce them fiercely. Again, it is only a matter of doing it smartly, without annoying otherwise sensible people.

**Comment by pauldreik** [comment]
I think we mostly agree on this, will come back to this. In the meanwhile, I link to an older existing issue: https://github.com/simdjson/simdjson/issues/798
